### 1번

In [1]:
fileS = open("s.txt",'r')
lines = []
while 1:
    line = fileS.readline().strip()
    if not line: break
    lines.append(line)
fileS.close()

lines.sort()
for i in range(len(lines)-1):
    lines[i] = lines[i]+'\n'

fileS1 = open("s1.txt",'w')
for line in lines:
    fileS1.write(line )
fileS1.close()

In [2]:
fileS = open("s.txt",'r')
linelist = []
while 1:
    line = fileS.readline().strip()
    if not line: break
    linelist.append(line.split(' '))
fileS.close()

linelist.sort(key = lambda x:x[1])
for i in range(len(linelist)-1):
    linelist[i] = linelist[i][0] + ' ' + linelist[i][1]+ '\n'
linelist[len(linelist)-1] = linelist[len(linelist)-1][0] + ' ' + linelist[len(linelist)-1][1]

fileS2 = open("s2.txt",'w')
for line in linelist:
    fileS2.writelines(line)
fileS2.close()

In [3]:
fileS = open("s.txt",'r')
linelist = []
while 1:
    line = fileS.read().replace('\n',' ')
    if not line: break
    linelist = line.split(' ')
fileS.close()

fileS3 = open("s3.txt",'w')
for i in range(len(linelist)):
    fileS3.write(linelist[i])
    if (i + 1) % 3 == 0:
        fileS3.write('\n')
    else:
        fileS3.write(' ')
fileS3.close()

1 의 1번은 작성하라는 내용이 없다. 코드작성을 안하는 것이라고 판단된다.
* 2)
> s.txt 를 r모드로 열고 빈 리스트(L)를 생성한다. L에 모든 줄을 줄넘김 문자를 제거하고, 추가한다. 그 뒤 L을 정렬한다.  
마지막 요소를 제외하고 리스트의 각 요소(string)에 줄넘김 문자를 추가한다.  
s1.txt 파일을 w모드로 열고 L의 모든 요소를 넣는다.
* 3)
> s.txt 를 r모드로 열고 빈 리스트(L)를 생성한다. L에 모든 줄에서 줄넘김 문자를 제거하고, 공백을 기준으로 분리한 리스트를 추가한다.  
L을 인덱스 1의 값으로 정렬, 즉 뒤의 단어를 기준으로 정렬한다.  
마지막 줄을 제외하고 리스트의 각 요소에 원본의 인덱스 0의 값, 공백, 인덱스 1의 값, 줄넘김문자를 합한 것을 대입한다.  
마지막 줄은 위와 똑같이 하되, 줄넘김 문자를 추가하지 않는다. s2.txt 파일을 w모드로 열고 L의 모든 요소를 넣는다.
* 4)
> s.txt 를 r모드로 열고 빈 리스트(L)를 생성한다. 텍스트를 모두 읽으며 줄넘김 문자를 공백으로 바꾸고, 공백을 기준으로 분리해 L에 넣는다.  
s3.txt 를 w모드로 열고, 각 요소를 집어넣는데, 집어넣은 횟수가 3이 될 때마다 줄넘김 문자를 추가해준다.

### 2번

In [1]:
import os
import hashlib

def enterID():
    ID = input('ID를 입력하세요. : ')
    if os.path.exists('access.txt'):
        file = open('access.txt', 'r')
        if os.stat('access.txt').st_size == 0:
            file.close()
            return ID
        while 1:
            line = file.readline().replace(',','').replace(':','')
            if not line: break
            lineList = line.split(' ')
            if ID == lineList[0]:
                print('Sorry, the enterd ID is already used')
                file.close()
                ID = enterID()
                break
        return ID
    else:
        return ID       

def enterPassword():
    Password = input('비밀번호를 입력하세요. : ')
    encoPassword = hashlib.sha1(Password.encode()).hexdigest()
    return encoPassword
    

def SignUp():
    print("계정 생성")
    
    ID = enterID()
    Password = enterPassword()
    Name = input('이름를 입력하세요. : ')
    School = input('학교를 입력하세요. : ')
    file = open('access.txt', 'a')
    if os.stat('access.txt').st_size == 0:
        file.write(ID + ': ' + Password + ', ' + Name + ', ' + School)
    else:
        file.write('\n' + ID + ': ' + Password + ', ' + Name + ', ' + School)
    file.close()

In [9]:
def checkID():
    enteredID = input("아이디를 입력하세요. : ")
    file = open('access.txt', 'r')
    while 1:
        line = file.readline().replace(',','').replace(':','')
        if not line: break
        lineList = line.split(' ')
        if enteredID == lineList[0]:
            file.close()
            return enteredID
    print('Sorry, you are not a registered member')
    file.close()
    checkID()
    
def checkPS(ID):
    password = enterPassword()   
    file = open('access.txt', 'r')
    while 1:
        line = file.readline().replace(',','').replace(':','')
        if not line: break
        lineList = line.split(' ')
        if ID == lineList[0]:
            if password == lineList[1]:
                print('Hello ' + lineList[2])
                file.close()
                return
            else:
                break
    print('Sorry, the entered password is not correct')
    file.close()
    checkPS(ID)

def SignIn():
    if os.path.exists('access.txt'):
        file = open('access.txt', 'r')
        if os.stat('access.txt').st_size == 0:
            print("작성된 계정정보가 없습니다.")
            file.close()
            return 0
    else:
        print("작성된 계정정보가 없습니다.")
        return 0
    
    print("로그인")
    ID = checkID()
    if ID:
        checkPS(ID)

In [11]:
print('Welcome to Our Service\n\tA. Sign Up\n\tB. Sign In\n\tC. Quit')
choice = int(input('선택하시오 : '))

if choice == 1:
    SignUp()
elif choice == 2:
    SignIn()
else:
    print('프로그램 종료')

Welcome to Our Service
	A. Sign Up
	B. Sign In
	C. Quit
선택하시오 : 3
프로그램 종료


* enterID()
> SignUp 함수에서 ID를 입력받는 용도로 쓰이는 함수다.  
ID를 입력받고, access.txt 파일이 존재하지 않거나, 빈 파일이면 ID를 반환한다.  
아니라면 파일의 각 줄을 읽는데, , 과 : 를 공백으로 치환시키고, 공백을 기준으로 분리한다. 파일 내 인덱스 0의 값이 ID이므로 일치하면 enterID를 다시 부른다. 모든 줄을 읽어서 없으면 ID를 반환한다.
* enterPassword()
> SignUp 함수에서 password를 입력받는 용도로 쓰이는 함수다.  
비밀번호를 입력받고 해시코드로 암호화한 뒤 반환한다.  
* SignUp()
> 계정생성을 담당하는 함수다.  
ID와 password는 enterID()와 enterPassword()의 반환값으로 입력받는다. 그리고 이름과 학교를 입력한다.  
access.txt 를 a모드로 열고 양식에 맞게 문자열을 만들어 넣는데, 앞에 줄바꿈 문자를 추가한다.  
다만 파일이 빈 경우는 줄바꿈 문자를 넣지 않는다.
* checkID()
> SignIn 함수에서 ID를 입력하고, 확인하는 용도로 쓰이는 함수다.  
ID를 입력받고 access.txt 파일을 r모드로 연다. enterID 함수처럼 읽어서 분리한 뒤 일치하는 ID가 있으면, ID를 반환한다. 만약 ID가 일치하지 않고 반복문을 나오면 메시지를 출력하고 checkID를 다시 호출한다.  
* checkPS()
> SignIn 함수에서 Password를 입력하고, 확인하는 용도로 쓰이는 함수다.  
읽어서 분리하는 부분까지는 checkID와 동일하다. 입력한 아이디와 읽은 아이디가 일치하며, 입력한 Password와 읽은 password가 일치하면 Hello 'name' 을 출력하고 함수를 나온다. 만약 Password가 모두 일치하지 않으면 메시지를 출력하고 checkPS를 다시 호출한다.


### 3번

In [34]:
def frange(val1, val2 = None, step = 0.1):
    if val2 == None:
        start = 0.0
        stop = val1
    else:
        start = val1
        stop = val2

    returnList = []
    while start <= stop:  
        returnList.append(start)
        start += step
        start = round(start,1)
    return returnList

frange(2.2, 4.0, 0.5)

[2.2, 2.7, 3.2, 3.7]

> 매개변수는 start stop step 순으로 입력되야 하지만 stop 만이 필수로 들어가야한다.  
그래서 val2를 None으로, step을 0.1으로 기본값 설정한 뒤, val2가 None인 경우는 매개변수를 하나만 받은 경우이므로, start를 0.0으로 초기화하고, stop을 val1값으로 한다. val2가 None이 아니라면 매개변수를 2개 이상 받은 경우이므로 start를 val1으로, stop을 val2로 초기화한다.  
리턴값을 줄 빈 리스트를 선언하고, start가 stop을 초과하지 않을 때까지 반복문을 돌리고 start에 step를 더하며 start를 리스트에 추가한다. 반복문이 끝나면 리스트를 반환한다

### 4번

In [12]:
def sum(*val):
    if len(val) == 0:
        return 0
    
    sumNumber = 0
    for i in val:
        sumNumber += i
    return sumNumber

sum(1,5,7,2,-10)

5

> 가변인수는 * 를 붙이는 방법으로 사용한다.  
가변인수(val)를 써서 함수를 생성하고 가변인수의 길이가 0이라면 입력값이 없기 때문에 0을 반환한다.  
0이 아니라면 합을 더할 변수 A를 0으로 초기화하고, val의 요소 각각을 뽑아 A에 더한 뒤 A를 반환한다.

### 5번

In [19]:
def strupper(strs):
    return strs[:1].upper() 

def myinitial(strs):
    strList = strs.split()
    strMap = list(map(strupper, strList))
    strJoin = "".join(strMap)
    print(strJoin)

myinitial("as soon as possible")

ASAP


* strupper(strs)
> myinitial 함수에서 쓰이는 함수다. 입력받은 strs에서 첫 자만 슬라이싱해서 빼내고, 대문자로 올린 문자열을 반환한다.
* myinitial(strs)
> strList에 입력받은 문자열을 공백을 구분자로 구분한다.  
map으로 strupper함수에 strList의 요소를 각각 넣고 list로 형변환해서 strMap에 대입한다.  
join으로 strMap요소를 빈 문자열에 집어넣고 해당 문자열을 출력한다.  

### 6번

In [17]:
def myfact(n):
    if n > 1:
        return n * myfact(n - 1)
    elif n == 0:
        return 0
    else:
        return 1

myfact(10)

3628800

> n이 1보다 크면 n보다 1 작은 수를 myfact 의 매개변수로 넣어서 나온 값과 n을 곱한 값을 반환한다.  
재귀를 통해 n이 점점 1씩 작아져서 n-1이 0인 경우 재귀가 종료된다.  
n이 0인 경우 0을 반환하고, n이 1인 경우 1을 반환한다.

### 7번

import string 의 경우 string 이라는 모듈 하나만 설치되어, 내부 모듈을 사용하려면 string.~~ 로 꺼내 써야 한다.  
반면 from strong import * 의 경우 string 내부의 모든 모듈이 각각 설치된다. string 이라는 모듈은 설치되지 않는다. 내부 모듈을 사용할 시 string. 을 쓸 필요가 없다.  

### Incremental

In [33]:
import requests
import string
import os
import pickle
    
# <script> 태그 제거
def cleanhtml_script(htmlText):
    j = 0;
    S = ''
    for i in range(len(htmlText)):
        i = htmlText.find('<script',j)
        S += htmlText[j:i]
        j = htmlText.find('</script>',i)
        if(j == -1):
            break
        j += 9               
    return S

# <style> 태그 제거
def cleanhtml_style(htmlText):
    j = 0;
    S = ''
    for i in range(len(htmlText)):
        i = htmlText.find('<style',j)
        S += htmlText[j:i]
        j = htmlText.find('</style>',i)
        if(j == -1):
            break
        j += 8         
    return S 

def cleanhtml_class(htmlText):
    j = 0;
    S = ''
    for i in range(len(htmlText)):
        i = htmlText.find('class=',j)
        S += htmlText[j:i]
        j = htmlText.find('>',i)
        if(j == -1):
            break  
    return S 

# 나머지 태그 제거
def cleanhtml_ex(htmlText):
    j = 0;
    S = ''
    for i in range(len(htmlText)):
        i = htmlText.find('>',j)
        if(i == -1 or i == len(htmlText)-1):
            break
        j = htmlText.find('<',i)
        if htmlText[i+1] != '<':    
            S += htmlText[i+1:j] + " "
    return S

# 태그제거 통합 & 리스트변환
def cleanhtml(htmlText):
    source1 = cleanhtml_script(htmlText)
    source2 = cleanhtml_style(source1)
    source3 = cleanhtml_ex(source2)
    source4 = cleanhtml_class(source3)
    textList = source4.split()
    return textList

# 구두문자 제거
def cleanpunct(textList):
    for i, text in enumerate(textList):
        for p in string.punctuation:
            if p in text:
                textList[i] = text.strip(string.punctuation)    
    
    return textList

# 공백요소 제거
def cleanblank(textList):
    cleanlist = ' '.join(textList).split()
    return cleanlist

# 딕셔너리화
def turndict(textList):
    textdict = {}
    for text in textList:
        if text not in textdict:
            textdict[text]=1
        else:
            textdict[text]+=1
            
    return textdict

# 내용 : 불용어 추출.
# 반환값 : 한글불용어, 영어불용어
# 사용처 : erasestopword() 함수 내 사용
def takestopword():
    file = open('stop_words_korean.txt','r', encoding = 'UTF8')
    LKoreanStopWord = []
    while 1:
        line = file.readline().strip()
        if not line: break
        splitline = line.split(' ')
        LKoreanStopWord.extend(splitline)
        
    file = open('stop_words_english.txt','r', encoding = 'UTF8')
    LEnglishStopWord = []
    while 1:
        line = file.readline().strip()
        if not line: break
        LEnglishStopWord.append(line)
    
    return LKoreanStopWord, LEnglishStopWord

# 입력값 : 소스 내 단어 딕셔너리(A)
# 내용 : A 내 불용어 삭제
# 출력값 : 불용어가 제거된 A
def erasestopword(DictSourceWord):
    LKoreanStopWord, LEnglishStopWord = takestopword()
    for dex in range(0, len(DictSourceWord)):    
        for stopword in LKoreanStopWord:
            if stopword in DictSourceWord:
                del DictSourceWord[stopword]
        for stopword in LEnglishStopWord:
            if stopword in DictSourceWord:
                del DictSourceWord[stopword]                                                
    return DictSourceWord

# 입력값 : Url
# 내용 : Url.html 파일 생성
def makehtmlfile(Url):
    if not os.path.exists('Incremental'):
        os.mkdir('Incremental')
    i = Url.find('://')   
    SUrlname = './Incremental/' + Url[i+3:] + '.html'
    file = open(SUrlname,'w')
    file.close()
        
# 입력값 : 불용어 제거된 딕셔너리, url 문자열
# 내용 : url.words_frequency 이름의 pickle 생성
def makepickle(DictSourceWord, Url):
    i = Url.find('://')   
    Urlname = './Incremental/' + Url[i+3:] + '.words_frequency.pickle'
    with open(Urlname,'wb') as fw:
        pickle.dump(DictSourceWord, fw)
        
# 입력값 : url 문자열
# 출력값 : 해당 url의 pickle 딕셔더리 데이터
def loadpickle(Url):
    i = Url.find('://')   
    Urlname = './Incremental/' + Url[i+3:] + '.words_frequency.pickle'
    with open(Urlname,'rb') as fr:
        sourcedict = pickle.load(fr)
    return sourcedict

# 입력값 : 불용어 제거된 딕셔너리
# 출력값 : 가장 많이 나온 단어 리스트 [1번째, 2번째, 3번째]
def picktopthree(DictSourceWord):
    Dsourceword = sorted(DictSourceWord.items(),reverse = True, key = lambda item: item[1])
    Ltopthree = [Dsourceword[0][0], Dsourceword[1][0], Dsourceword[2][0]]
    return Ltopthree

# 출력값 : 찾을 단어 리스트
def getsearchword():
    Sword = input('검색어를 입력하세요. : ')
    Lword = Sword.split()
    LKoreanStopWord, LEnglishStopWord = takestopword()    
    for stopword in LKoreanStopWord:
        if stopword in Lword:
            Lword.remove(stopword)
    for stopword in LEnglishStopWord:
        if stopword in Lword:
            Lword.remove(stopword)
    return Lword

# 입력값 : 찾을 단어 리스트, url단어 리스트
# 출력값 : url 단어리스트 속 찾는 단어:개수 와 총 개수가 입력된 딕셔너리
def foundword(Lword, Durlword):
    Dword = {}
    for word in Lword:
        if word in Durlword:
            Dword[word] = Durlword.get(word)
    total = 0
    for count in Dword.values():
        total += count
    Dword['Total'] = total
    return Dword

# 입력값 : url 리스트, 찾을 단어 리스트, url 단어 딕셔너리
# 내용 : 찾은 모든 단어 수가 가장 많은 것을 기준으로 url을 내림차순 출력
def checksimilarity(Lurl, Lword, Durlword):
    totalwords = {}
    for i in range(0,5):
        totalwords[Lurl[i]] = foundword(Lword, Durlword[i]).get('Total')
    totalwords = sorted(totalwords.items(),reverse = True, key = lambda item: item[1])
    for i in range(0,5):
        print(totalwords[i][0])

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

# 태그, 구두문자, 공백 제거. + 딕셔너리 변환
def integratedmodule_eraseddict(source):
    S1 = cleanhtml(source)
    S2 = cleanpunct(S1)
    S3 = cleanblank(S2)
    S4 = turndict(S3)
    S5 = erasestopword(S4)
    return S5

# 디버깅 용 url 입력값
LUrl = ['http://cse.koreatech.ac.kr', 'https://www.koreatech.ac.kr', 'http://www.naver.com',
        'http://www.daum.net', 'http://www.nytimes.com']

# LUrl = []
Lsource_dict = []
Lsource_load_dict = []

# 1번
for i in range(0,5):
    # LUrl.append(input('URL을 입력하세요. : '))
    req = requests.get(LUrl[i])
    Lsource_dict.append(integratedmodule_eraseddict(req.text))  

# 2번, 3번, 4번, 5번
for i in range(0,5):
    makehtmlfile(LUrl[i])
    makepickle(Lsource_dict[i],LUrl[i])
    print(picktopthree(loadpickle(LUrl[i])))

# 6번
Lword = getsearchword()
checksimilarity(LUrl, Lword, Lsource_dict)


['Page', 'Error', '404']
['한국기술교육대학교', '대학', '제공합니다']
['리빙', '전', '구독']
['날씨', '°C', '맑음']
['The', 'New', 'Times']
검색어를 입력하세요. : 한국기술교육대학교 에서 대학을 맑음 네이버 에 가 New
http://www.nytimes.com
http://www.daum.net
https://www.koreatech.ac.kr
http://www.naver.com
http://cse.koreatech.ac.kr


* takestopword()  
> 한글 불용어의 uml을 들어갈 시 404 NOT FOUND 가 나와서, txt파일로 만들었다.  
한줄씩 읽어, 줄바꿈글자를 제거하고 빈 리스트에 넣었다. 그리고 반환한다.   
다만 한글은 띄어쓰기가 있는 경우 분리를 하기 위해 split으로 분리해서 리스트를 따로 생성하고, extend를 써서 이어 붙이는 방식으로 추가했다.  
* erasestopword(A)  
> takestopword 를 통해 불용어를 받고, 불용어 리스트에서 단어를 하나씩 뽑아 A에 존재하는 지 확인하고, 있다면 제거했다. 그리고 A를 반환한다.  
* makehtmlfile(url)  
> Incremental 디렉토리가 있는 지 확인하고, 없으면 생성한다. string형식으로 된 url에서 :// 의 인덱스를 찾는다.  
://의 글자 수인 3을 찾은 인덱스에 더해서 그 뒤로 끝까지 슬라이싱하고, 앞에 ./Incremental/ 을, 뒤에 .html 을 붙인다.  
그렇게 만든 문자열의 이름으로 파일을 w모드로 열고 닫아 생성하면, Incremental 디렉토리 안에 html 파일이 생성된다.  
* makepickle(DickSourceWord, Url)
> makehtmlfile()과 동일하게 하되, 끝에 .html이 아닌 .words_frequency.pickle 을 붙인다.  
만들어진 문자열을 wb모드로 열고, 딕셔너리를 dump로 저장한다.  
* loadpickle(Url)
> makepickle과 문자열 생성까지는 동일하다. 문자열을 rb모드로 열고 내용을 불러온 뒤 반환한다.
* picktopthree(DictSourceWord)
> DictSourceWord를 값기준, 역순으로 정렬하고, 0 부터 2번지에 해당하는 키를 리스트로 모아 반환한다.
* getsearchword()
> 검색어를 입력받아, 공백을 기준으로 나눠 리스트화 시킨다. erasestopword()와 유사한 반복문을 걸처 단어 리스트에서 불용어를 제거하고 반환한다.
* foundword(Lword, Durlword)
> 빈 딕셔너리(D)를 선언한다. 찾을 단어리스트에서 단어를 하나씩 for문으로 가져와서 url 단어리스트에 존재하면 D에 해당 단어와 갯수를 추가한다.  그 후 각 값을 하나씩 뽑아서 더해 총 개수를 구해서 Total이라는 키로 저장한다.
* checksimilarity(Lurl, Lword, Durlword)
> 빈 딕셔너리(D)를 선언한다. 0부터 5까지의 인덱스를 for문으로 돌려 D에 각 인덱스에 해당하는 url를 키로, foundword() 함수를 돌려서 나온 딕셔너리에서 키'Total'의 값을 값으로 넣는다. 그리고 값을 기준으로 내림차순 정렬시킨다.   
반복문을 돌려서 url부분을 순서대로 출력하면, 찾은 모든 단어 수가 가장 많은 기준 내림차순으로 나온다.
* 추가점
> 지금 구현된 검색엔진은 모든 단어가 나온 횟수를 더해서, 그 수가 가장 큰 값이 유사성이 높다고 판단했다.  
이는 단어가 아니라 문장을 검색할 경우 각각의 단어가 각자 다른 곳에서 검사되기 때문에 문장으로서의 유사성은 없다. 문장을 구분자로 분리하지 않을 때의 검색이 되야한다. 

2018136014_김상우
frange 함수를 작성할 때 range의 형태 중 start가 디폴트 값이 있는 인자인데, 어떻게 앞에 있는지 정말 궁금하다.